Workflows 9.4 content — part A (#6048)
elastic/docs-content: explore-analyze/workflows/steps/cases.md

---
navigation_title: Cases
applies_to:
  stack: ga 9.4
  serverless: ga
description: Reference for the 27 Cases action steps that let workflows create, query, update, attach content, and manage the lifecycle of Elastic Security cases.
products:
  - id: kibana
  - id: cloud-serverless
  - id: cloud-hosted
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: elastic-stack
---

# Cases action steps [workflows-cases-steps]

> Cases action steps let workflows create, query, update, and manage cases in {{elastic-sec}} and other Cases-enabled apps. The `cases.*` namespace provides full coverage of the Cases API so you can automate case workflows without leaving the Elastic platform.

Reviewer: Is this emphasis on security here necessary?

Use Cases steps for patterns like:

- Create a case when a detection rule fires, then attach the alert and relevant observables.
- Open a case, assign it to an on-call user, and post a notification to a team channel.
- Find similar cases for deduplication before creating a new one.
- Close a case automatically when a condition is met, and append an analyst note first.

## Shared conventions [workflows-cases-conventions]

Every `cases.*` step shares the same conventions, so once you learn one step the others are predictable.

**Parameter casing.** Case-level parameters use `snake_case`: `case_id`, not `caseId`. The editor rejects camelCase for these fields. (Individual payloads, such as the alert object attached by `cases.addAlerts`, use the casing of the originating API (for example, `alertId`, `index`) because they match external API shapes.)

> **Optional `push-case` config.** All 27 steps accept an optional top-level `push-case` boolean (defaults to `false`). When `true`, the step pushes the updated case to a connected external system (for example, {{jira}} or ServiceNow) after the workflow operation succeeds.

Reviewer: Not all cases steps support `push-case`

Author: Do you have a list of which ones support it?

Reviewer: It's all steps that perform changes on the case. So it's not supported by deleteCase, findCases, findSimilarCases, getAllAttachments, getCase, getCases, getCasesByAlertId. Also not by updateObservable because that's just internal.

```yaml
- name: mark_in_progress
  type: cases.setStatus
  with:
    case_id: "{{ steps.create_case.output.case.id }}"
    status: "in-progress"
```

### `cases.setSeverity` [cases-setseverity]

Set a case's severity.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `severity` | `with` | string | Yes | `low`, `medium`, `high`, or `critical`. |

### `cases.setTitle` [cases-settitle]

Update a case's title.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `title` | `with` | string | Yes | New title. |

### `cases.setDescription` [cases-setdescription]

Update a case's description.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `description` | `with` | string | Yes | New description. Markdown supported. |

### `cases.setCategory` [cases-setcategory]

Set a case's category.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |

> | `category` | `with` | string | Yes | Category name. |

Reviewer: `owner` is another optional parameter

### `cases.setCustomField` [cases-setcustomfield]

Set a single custom-field value on a case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `field_name` | `with` | string | Yes | Custom field identifier. |
| `value` | `with` | `string \| number \| boolean` | Yes | Field value. |

> | `version` | `with` | string | No | Case version for optimistic concurrency. |

Reviewer: Same here, `owner` is optional and when provided helps with auto-completion

```yaml
- name: set_investigation_owner
  type: cases.setCustomField
  with:
    case_id: "{{ steps.create_case.output.case.id }}"

>     field_name: "investigation_owner"

Reviewer: `field_name` is actually a system-set ID, maybe we can change this example to look more like an ID?

Author: I changed it to a UUID style example. If there's another format we should use lmk

|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |

Output: `{ attachments: array }`.

```yaml
- name: list_attachments
  type: cases.getAllAttachments
  with:
    case_id: "{{ inputs.case_id }}"
```

---

## Tags and assignees

### `cases.addTags` [cases-addtags]

Add tags to a case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `tags` | `with` | `string[]` | Yes | Tags to add. |

### `cases.assignCase` [cases-assigncase]

Assign a case to one or more users.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `assignees` | `with` | array | Yes | Array of `{ uid }` objects. |

```yaml
- name: assign_oncall
  type: cases.assignCase
  with:
    case_id: "{{ steps.create_case.output.case.id }}"
    assignees:
      - uid: "{{ consts.oncall_uid }}"
```

### `cases.unassignCase` [cases-unassigncase]

Remove assignees from a case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |

> | `assignees` | `with` | array | Yes | Array of `{ uid }` objects to remove. |

Reviewer: If an empty array is passed here, all assignees will be removed

---
navigation_title: Cases
applies_to:
  stack: ga 9.4
  serverless: ga

> description: Reference for the 27 Cases action steps that let workflows create, query, update, attach content, and manage the lifecycle of cases in any Cases-enabled app.

Reviewer 2: Minor tweak to future-proof the description. 

```suggestion
description: Reference for the Cases action steps that let workflows create, query, update, attach content, and manage the lifecycle of cases in Kibana, Observability, and Security.
```

products:
  - id: kibana
  - id: cloud-serverless
  - id: cloud-hosted
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: elastic-stack
---

# Cases action steps [workflows-cases-steps]

Cases action steps let workflows create, query, update, and manage cases. The `cases.*` namespace provides full coverage of the Cases API through a single, consistent set of step types so you can automate case workflows without leaving the Elastic platform.

Use Cases steps for patterns like:

- Create a case when a detection rule fires, then attach the alert and relevant observables.
- Open a case, assign it to an on-call user, and post a notification to a team channel.
- Find similar cases for deduplication before creating a new one.
- Close a case automatically when a condition is met, and append an analyst note first.

## Shared conventions [workflows-cases-conventions]

Every `cases.*` step shares the same conventions, so once you learn one step the others are predictable.

> **Parameter casing.** Case-level parameters use `snake_case`: `case_id`, not `caseId`. The editor rejects camelCase for these fields. (Individual payloads, such as the alert object attached by `cases.addAlerts`, use the casing of the originating API (for example, `alertId`, `index`) because they match external API shapes.)

Reviewer 2: ```suggestion
**Parameter casing.** Case-level parameters use `snake_case`: `case_id`, not `caseId`. The editor rejects camelCase for these fields. Individual payloads, such as the alert object attached by `cases.addAlerts`, use the casing of the originating API (for example, `alertId`, `index`) because they match external API shapes.
```

**Optional `push-case` config.** Many `cases.*` steps accept an optional top-level `push-case` boolean (defaults to `false`). When `true`, the step pushes the updated case to a connected external system (for example, {{jira}} or ServiceNow) after the workflow operation succeeds. Refer to each step's parameters for whether `push-case` is supported.

```yaml
- name: create_and_push
  type: cases.createCase
  push-case: true
  with:
    title: "Triage required"
    severity: high
```

**Getting the case ID.** The response from `cases.createCase` includes the new case ID at `steps.<step_name>.output.case.id`. Reference it in subsequent steps:

```yaml
- name: new_case
  type: cases.createCase
  with:
    title: "..."

- name: tag_case
  type: cases.addTags
  with:
    case_id: "{{ steps.new_case.output.case.id }}"
    tags: ["triaged"]
```

## Step catalog [workflows-cases-catalog]

The 27 Cases steps group into six operational categories. Jump to any step:

> 

Reviewer 2: I don't think we have style guidelines for this kind of format. A list would be more scannable if you're still making changes to this PR : )

```

### `cases.getCase` [cases-getcase]

Retrieve a complete case object by ID. Optionally include comments and attachments in the response.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `include_comments` | `with` | boolean | Yes | Whether to include comments and attachments in the response. |

```yaml
- name: fetch_case
  type: cases.getCase
  with:
    case_id: "{{ inputs.case_id }}"
    include_comments: true
```

### `cases.getCases` [cases-getcases]

Retrieve up to 1000 cases in a single request. IDs that can't be fetched are reported in the `errors` array on the output. Use this to avoid N sequential `cases.getCase` calls in fan-out workflows.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_ids` | `with` | `string[]` | Yes | Array of case IDs. Maximum 1000. |

Output: `{ cases: array, errors: array }`. The `errors` array contains entries for any IDs that couldn't be fetched.

```yaml
- name: fetch_cases
  type: cases.getCases
  with:
    case_ids: ["case-1", "case-2", "case-3"]
```

### `cases.findCases` [cases-findcases]

> Search for cases matching filter criteria. Supports paging, sorting, and filtering by many fields.

Reviewer 2: ```suggestion
Search for cases matching filter criteria. Supports paging, sorting, and filtering by many fields. The output includes a `cases` array plus per-status counts (`count_open_cases`, `count_in_progress_cases`, `count_closed_cases`), `page`, `per_page`, and `total`.
```

| `tags` | `with` | `string \| string[]` | No | Filter by tag. |
| `category` | `with` | `string \| string[]` | No | Filter by category. |
| `owner` | `with` | `string \| string[]` | No | Filter by case owner. |
| `assignees` | `with` | `string \| string[]` | No | Filter by assignee UID. |
| `reporters` | `with` | `string \| string[]` | No | Filter by reporter. |
| `customFields` | `with` | object | No | Map of custom-field ID to allowed values. |
| `from` | `with` | string | No | Start of the time range. |
| `to` | `with` | string | No | End of the time range. |
| `page` | `with` | number | No | Page number. Defaults to `1`. |
| `perPage` | `with` | number | No | Results per page. Defaults to `20`. |
| `sortField` | `with` | string | No | `title`, `category`, `createdAt`, `updatedAt`, `closedAt`, `status`, or `severity`. |
| `sortOrder` | `with` | string | No | `asc` or `desc`. |

```yaml
- name: find_open_high_sev
  type: cases.findCases
  with:
    owner: "securitySolution"
    status: "open"
    severity: ["high", "critical"]
    tags: ["investigation"]
    sortField: "updatedAt"
    sortOrder: "desc"
    perPage: 20
```

> The output includes a `cases` array plus per-status counts (`count_open_cases`, `count_in_progress_cases`, `count_closed_cases`), `page`, `per_page`, and `total`.

Reviewer 2: Is this about the example directly before this line, or is it a general statement about the output of the `cases.findCases` step? If it's the latter, I'd recommend [moving it to above the table](https://github.com/elastic/docs-content/pull/6048/changes#r3157766264). 

```suggestion
```

### `cases.findSimilarCases` [cases-findsimilarcases]

Find cases similar to a given case, matched by shared observables. Useful for deduplication before creating a new case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Source case ID to find similar cases for. |
| `page` | `with` | integer | Yes | Page number (1-based). |
| `perPage` | `with` | integer | Yes | Results per page. |

> Output: `{ cases: array, page: integer, per_page: integer, total: integer }`.

Reviewer 2: ```suggestion
The output includes a `cases` array, `page`, and `per_page`.
```

| `events` | `with` | array | Yes | Array of `{ eventId, index }` objects. |

```yaml
- name: attach_context_events
  type: cases.addEvents
  with:
    case_id: "{{ steps.create_case.output.case.id }}"
    events:
      - eventId: "{{ steps.search_context.output.hits.hits[0]._id }}"
        index: "{{ steps.search_context.output.hits.hits[0]._index }}"
```

### `cases.addObservables` [cases-addobservables]

Add observables (indicators of compromise such as IPs, file hashes, domains, or URLs) to a case. `typeKey` must match one of the built-in observable types in {{elastic-sec}}.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `observables` | `with` | array | Yes | Array of `{ typeKey, value, description? }` objects. |

```yaml
- name: add_iocs
  type: cases.addObservables
  with:
    case_id: "{{ steps.create_case.output.case.id }}"
    observables:
      - typeKey: "observable-type-ipv4"
        value: "{{ event.alerts[0].source.ip }}"
        description: "Source of malicious activity"
      - typeKey: "observable-type-hash-sha256"
        value: "{{ event.alerts[0].file.hash.sha256 }}"
```

> The `typeKey` must match one of the built-in observable type keys (for example, `observable-type-ipv4`, `observable-type-ipv6`, `observable-type-url`, `observable-type-domain`, `observable-type-hash-sha256`, `observable-type-hash-md5`).

Reviewer 2: Optional edit to break up this v long sentence:

```suggestion
The `typeKey` must match one of the built-in observable type keys. Examples of accepted values include: `observable-type-ipv4`, `observable-type-ipv6`, `observable-type-url`, `observable-type-domain`, `observable-type-hash-sha256`, `observable-type-hash-md5`
```

### `cases.getAllAttachments` [cases-getallattachments]

Fetch every attachment associated with a case without pagination. Use this when you need the complete set of attachments for decision-making, for example when checking evidence before closing or escalating.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |

> Output: `{ attachments: array }`.

Reviewer 2: ```suggestion
The output is a list of attachments associated with the case you specified.
```

---

## Update and delete observables

### `cases.updateObservable` [cases-updateobservable]

Update an existing observable on a case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `observable_id` | `with` | string | Yes | Observable ID. |
| `value` | `with` | string | Yes | Updated value. |
| `description` | `with` | `string \| null` | No | Updated description. |

### `cases.deleteObservable` [cases-deleteobservable]

Remove an observable from a case.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_id` | `with` | string | Yes | Case ID. |
| `observable_id` | `with` | string | Yes | Observable ID. |

---

## Delete

### `cases.deleteCases` [cases-deletecases]

Delete one or more cases permanently, including their attachments and comments.

| Parameter | Location | Type | Required | Description |
|---|---|---|---|---|
| `case_ids` | `with` | `string[]` | Yes | Array of case IDs to delete. |

```yaml
- name: cleanup
  type: cases.deleteCases
  with:
    case_ids: ["old-case-1", "old-case-2"]
```

:::{warning}

> Deleted cases can't be recovered. Prefer [`cases.closeCase`](#cases-closecase) unless you really need the case removed.

Reviewer 2: Optional edit to put the reasoning first.

```suggestion
Deleted cases can't be recovered. If you don't want to permanently delete a case, use [`cases.closeCase`](#cases-closecase) instead.
```

:::

## Migrating from `kibana.*` case aliases [workflows-cases-migration]

In 9.3, case management lived under the `kibana.*` namespace. Those step types remain as deprecated aliases so existing workflows keep running, but new workflows must use `cases.*`:

| Deprecated (9.3 alias) | Current (9.4) |
|---|---|
| `kibana.createCaseDefaultSpace` | [`cases.createCase`](#cases-createcase) |
| `kibana.getCaseDefaultSpace` | [`cases.getCase`](#cases-getcase) |
| `kibana.updateCaseDefaultSpace` | [`cases.updateCase`](#cases-updatecase) |
| `kibana.addCaseCommentDefaultSpace` | [`cases.addComment`](#cases-addcomment) |

> See [Migrate from 9.3 to 9.4](/explore-analyze/workflows/authoring-techniques/migrate-from-9.3.md) for side-by-side replacement patterns.

Reviewer 2: ```suggestion
Refer to [Migrate from 9.3 to 9.4](/explore-analyze/workflows/authoring-techniques/migrate-from-9.3.md) for side-by-side replacement patterns.
```

Reviewer 2: Also, I'm not sure why the [Migrate from 9.3 to 9.4](https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/6048/explore-analyze/workflows/authoring-techniques/migrate-from-9.3) ref on line 626 is 404'ing. Looks like the same ref on a different page ([Migrate workflows from 9.3 to 9.4](https://docs-v3-preview.elastic.dev/elastic/docs-content/pull/6048/explore-analyze/workflows/authoring-techniques/migrate-from-9.3)) is 404ing too. : (
